### PR TITLE
Tag data corruption errors

### LINKED ISSUE #4663: Data Corruption Error Categorization
There are a number of cases where the an error represents data corruption, we should ensure all these cases throw errors that indicate they are data corruption.

### PR DESCRIPTION
**Pull request by jcmoore**
@anthony-murphy , looking for advice where `extractLogSafeMessageProperties` (and `extractLogSafeErrorProperties`) ought to be moved and tested.

### CODE REVIEW

**Comment by anthony-murphy** [comment]
probably @fluidframework/telemetry-utils

File: packages/dds/merge-tree/src/mergeTree.ts [hunk 1]
```diff
@@ -2226,11 +2227,11 @@ export class MergeTree {
                     { leaf: onLeaf, candidateSegment: newSegment, continuePredicate: continueFrom });
 
                 if (newSegment.parent === undefined) {
-                    throw new Error(`MergeTree insert failed: ${JSON.stringify({
-                        currentSeq: this.collabWindow.currentSeq,
-                        minSeq: this.collabWindow.minSeq,
-                        segSeq: newSegment.seq,
-                    })}`);
+                    throw new DataCorruptionError("MergeTree insert failed", {
+                        referenceSequenceNumber: this.collabWindow.currentSeq,
```
Comment (anthony-murphy) [inline packages/dds/merge-tree/src/mergeTree.ts:2231]: this isn't referenceSequenceNumber, that's refSeq. currentSeq is the last seq seen by the merge tree

Comment (anthony-murphy) [reply packages/dds/merge-tree/src/mergeTree.ts:2231]: would be good to have both

File: packages/loader/container-loader/src/deltaManager.ts [hunk 2]
```diff
@@ -1359,10 +1359,20 @@ export class DeltaManager
         }
 
         // Watch the minimum sequence number and be ready to update as needed
-        assert(this.minSequenceNumber <= message.minimumSequenceNumber, "msn moves backwards");
+        if (this.minSequenceNumber > message.minimumSequenceNumber) {
+            throw new DataCorruptionError("msn moves backwards", {
+                ...extractLogSafeMessageProperties(message),
+                ...(this.connection && { clientId: this.connection?.clientId}),
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/deltaManager.ts:1365]:
```suggestion
                clientId: this.connection?.clientId,
```
minor simplification

File: packages/loader/container-loader/src/deltaManager.ts [hunk 3]
```diff
@@ -1359,10 +1359,20 @@ export class DeltaManager
...
+            });
+        }
         this.minSequenceNumber = message.minimumSequenceNumber;
 
-        assert(message.sequenceNumber === this.lastProcessedSequenceNumber + 1, "non-seq seq#");
+        if (message.sequenceNumber !== this.lastProcessedSequenceNumber + 1) {
+            throw new DataCorruptionError("non-seq seq#", {
+                ...extractLogSafeMessageProperties(message),
+                ...(this.connection && { clientId: this.connection?.clientId}),
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/deltaManager.ts:1373]:
```suggestion
                clientId: this.connection?.clientId,
```

File: packages/loader/container-loader/src/deltaManager.ts (hunk 2 above)
Comment (jcmoore) [reply packages/loader/container-loader/src/deltaManager.ts:1365]: If there is no concern about `undefined` "pollution"... I'll probably reimplement [`extractLogSafeMessageProperties`](https://github.com/microsoft/FluidFramework/pull/5188/files#diff-01eeb7567dd98442afb9a308001150972f600ef3f04e17a71f88b34a0b8ee153R1462) in a simpler fashion as well

File: packages/dds/merge-tree/src/mergeTree.ts (hunk 1 above)
Comment (jcmoore) [reply packages/dds/merge-tree/src/mergeTree.ts:2231]: CollaborationWindow here only has `currentSeq`, `minSeq`, and `localSeq` as far as I can tell.  Still not sure I'm parsing the nuances of "current" and "local".  Were you suggesting refSeq might come from somewhere else entirely?

Comment (anthony-murphy) [reply packages/dds/merge-tree/src/mergeTree.ts:2231]: refSeq is a parameter to this function, so the refSeq of the op for this insert. collab window is the state of the merge tree itself, so current is the last seq that's ever been see by this object across all ops

File: packages/loader/container-loader/src/deltaManager.ts (hunk 2 above)
Comment (anthony-murphy) [reply packages/loader/container-loader/src/deltaManager.ts]: i'm not particularly concerned as these objects are short lived, and only created in error cases. if this was on some critical code path we might need to worry

File: packages/dds/merge-tree/package.json [hunk 4]
```diff
@@ -56,6 +56,7 @@
     "@fluidframework/common-definitions": "^0.19.1",
     "@fluidframework/common-utils": "^0.27.0",
     "@fluidframework/container-definitions": "^0.35.0",
+    "@fluidframework/container-utils": "^0.35.0",
```
Comment (anthony-murphy) [inline packages/dds/merge-tree/package.json:59]: ok. it looks like this is allowed, but i'm not sure it should be. i'd prefer to move the error to telemetry utils, which is a bigger change

**Review by anthony-murphy — APPROVED** [review]

File: packages/loader/container-loader/src/deltaManager.ts [hunk 5]
```diff
@@ -1424,10 +1424,20 @@ export class DeltaManager
         }
 
         // Watch the minimum sequence number and be ready to update as needed
-        assert(this.minSequenceNumber <= message.minimumSequenceNumber, "msn moves backwards");
+        if (this.minSequenceNumber > message.minimumSequenceNumber) {
```
Comment (vladsud) [inline packages/loader/container-loader/src/deltaManager.ts:1427]: How about we create assertFluidOpInvariant() or some other name that is less verbose and achieves same thing?
There is no need to add clientID - it's already in telemetry and easy to find

Comment (jcmoore) [reply packages/loader/container-loader/src/deltaManager.ts:1427]: (Much more than one might suspect) I really appreciate you offering up a name @vladsud !  Where would be a reasonable place to `export` it (if it is not file local)?  Would packages/utils/telemetry-utils/src/assert.ts (file does not yet exist) make sense?

Comment (vladsud) [reply packages/loader/container-loader/src/deltaManager.ts:1427]: telemetry-utils is generic telemetry package (pretty much not fluid specific), it should not take dependency on protocol or other packages. So I think either driver-utils or container-utils (or maybe new package?)

**Comment by anthony-murphy** [comment]
even thought his has some feedback i'm going to commit this, as i'd like the error categorized. we can improve in any followups

**Pull request merged**